flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java

```diff
@@ -188,27 +193,39 @@ public StreamGraphGenerator setJobName(String jobName) {
 	}
 
 	public StreamGraphGenerator setGlobalDataExchangeMode(GlobalDataExchangeMode globalDataExchangeMode) {
-		this.globalDataExchangeMode = globalDataExchangeMode;
+		this.globalDataExchangeMode = checkNotNull(globalDataExchangeMode);
 		return this;
 	}
 
-	public void setSavepointRestoreSettings(SavepointRestoreSettings savepointRestoreSettings) {
+	public StreamGraphGenerator setSavepointRestoreSettings(SavepointRestoreSettings savepointRestoreSettings) {
 		this.savepointRestoreSettings = savepointRestoreSettings;
+		return this;
+	}
+
+	public StreamGraphGenerator setAllVerticesInSameSlotSharingGroupByDefault(boolean allVerticesInSameSlotSharingGroup) {
+		this.allVerticesInSameSlotSharingGroup = allVerticesInSameSlotSharingGroup;
+		return this;
 	}
 
 	public StreamGraph generate() {
-		streamGraph = new StreamGraph(executionConfig, checkpointConfig, savepointRestoreSettings);
+		GlobalDataExchangeMode finalDataExchangeMode = determineGlobalDataExchangeMode();
+		ScheduleMode finalScheduleMode = determineScheduleMode(finalDataExchangeMode);
+		streamGraph = new StreamGraph(
+			executionConfig,
+			checkpointConfig,
+			savepointRestoreSettings,
+			finalDataExchangeMode);
```

Comment:
Most of the following setters can become constructor arguments, right? This seems to be the only place they are called.